src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -189,16 +204,51 @@ private async Task<AuthenticationResult> SendTokenRequestToAppTokenProviderAsync
 
         private async Task<MsalAccessTokenCacheItem> GetCachedAccessTokenAsync()
         {
-            MsalAccessTokenCacheItem cachedAccessTokenItem = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
+            // Fetch the cache item (could be null if none found).
+            MsalAccessTokenCacheItem cacheItem =
+                await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
 
-            if (cachedAccessTokenItem != null && !_clientParameters.ForceRefresh)
+            // If the item fails any checks (null, or hash mismatch),
+            if (!ShouldUseCachedToken(cacheItem))
             {
-                AuthenticationRequestParameters.RequestContext.ApiEvent.IsAccessTokenCacheHit = true;
-                Metrics.IncrementTotalAccessTokensFromCache();
-                return cachedAccessTokenItem;
+                return null;
             }
 
-            return null;
+            // Otherwise, record a successful cache hit and return the token.
+            MarkAccessTokenAsCacheHit();
+            return cacheItem;
+        }
+
+        private bool ShouldUseCachedToken(MsalAccessTokenCacheItem cacheItem)
+        {
+            // 1) No cached item 
+            if (cacheItem == null)
+            {
+                return false;
+            }
+
+            // 2) If the token’s hash matches AccessTokenHashToRefresh, ignore it
+            if (!string.IsNullOrEmpty(_clientParameters.AccessTokenHashToRefresh) &&
+                IsMatchingTokenHash(cacheItem.Secret, _clientParameters.AccessTokenHashToRefresh))
+            {
+                AuthenticationRequestParameters.RequestContext.Logger.Info(
+                    "[ClientCredentialRequest] A cached token was found and its hash matches AccessTokenHashToRefresh, so it is ignored.");
```

Comment:
Given that we now seem to lean towards only implementing this in MSAL .Net, shall we add that [telemetry "TokenRejected"](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/5179#discussion_r1987572831) to MSAL .Net, too, preferably in this PR?

Reply:
Sure @rayluo

Reply:
skipping for now